bin/generate_protein_and_entity_ids.py

```diff
@@ -153,36 +142,21 @@ def main(args=None):
                         ]:
                             entities = pd.DataFrame()
                             entities = proteins[["entity_name"]].drop_duplicates()
-                            entities["entity_id"] = range(next_entity_id, next_entity_id + len(entities))
+
                             # Instead of microbiome_bare_id append microbiome_id
                             entities["microbiome_id"] = microbiome_id
                             all_entities.append(entities)
 
-                        next_entity_id += len(entities)
                         entities = pd.concat(all_entities)
 
                     else:
                         entities = proteins[["entity_name"]].drop_duplicates()
-                        entities["entity_id"] = range(next_entity_id, next_entity_id + len(entities))
-                        next_entity_id += len(entities)
                         # If not coassembled microbiome_id = microbiome_bare_id
                         entities["microbiome_id"] = microbiome_bare_id
 
-                    # Write proteins
-                    proteins.rename(columns={"protein_tmp_id": "protein_orig_id"}, inplace=True)
-                    proteins[proteins_columns].to_csv(outfile_proteins, sep="\t", header=False, index=False)
-                    # Write entities_proteins
-                    proteins.merge(entities)[["entity_id", "protein_id"]].drop_duplicates().to_csv(
-                        outfile_entities_proteins, sep="\t", header=False, index=False
-                    )
-                    # Write entities
-                    entities[entities_columns].drop_duplicates().to_csv(
-                        outfile_entities, sep="\t", header=False, index=False
-                    )
-                    # Write microbiomes - entities
-                    entities[microbiomes_entities_columns].to_csv(
-                        outfile_microbiomes_entities, sep="\t", index=False, header=False
-                    )
+            # Merge all dfs of all microbiomes
+                proteins_dfs.append(proteins)
+                entities_dfs.append(entities)
```

Comment:
Is the issue here that the protein order differs? Similar as in `generate_peptides.py`, and could this maybe changed at the protein level already?

Reply:
Yes probably True for the `PRODIGAL` process output, but I'm not so sure about the `ENTREZ` downloads! This way both is merged, sorted at one place so that the protein ids can be given here. Do you think I should try to solve it differently? I can check again to see if I'm able to get a reproducible output if I sort the input/output of PRODIGAL

Reply:
I would try to avoid to many code changes, and try to solve the problem directly where it comes from, this would also keep the code ore intuitively

Reply:
Generally yes, and I think it's worth to check again! Still I would change the script similar, as it is not easily readable. The data frames are continuously written to files, both for bins/assembly and for download. This creates a lot of redundant code, and imo obscures the actual output of the script. I think it's easier to understand if all data is collected, sorted and then ids are assigned and output files generated, we can discuss on Monday when I checked the output of prodigal again! 🙂